analytic_engine/src/memtable/columnar/mod.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic,
+        atomic::{AtomicU64, AtomicUsize, Ordering},
+        Arc, RwLock,
+    },
+};
+
+use arena::MonoIncArena;
+use bytes::Bytes;
+use common_types::{column::Column, datum::Datum, row::Row, schema::Schema, SequenceNumber};
+use common_util::error::BoxError;
+use log::debug;
+use skiplist::Skiplist;
+use snafu::{ensure, OptionExt, ResultExt};
+
+use crate::memtable::{
+    columnar::iter::ColumnarIterImpl,
+    factory::Options,
+    iter::ReversedColumnarIterator,
+    key::{BytewiseComparator, KeySequence},
+    ColumnarIterPtr, Internal, InternalNoCause, InvalidPutSequence, MemTable, PutContext, Result,
+    ScanContext, ScanRequest,
+};
+
+pub mod factory;
+pub mod iter;
+
+pub struct ColumnarMemTable {
+    /// Schema of this memtable, is immutable.
+    schema: Schema,
+    memtable: Arc<RwLock<HashMap<String, Column>>>,
+    /// The last sequence of the rows in this memtable. Update to this field
+    /// require external synchronization.
+    last_sequence: AtomicU64,
+    row_num: AtomicUsize,
+    opts: Options,
```

Comment:
Better to avoid save the whole `Options` here, I suggest only keep required fields.